calico_kubernetes/calico_kubernetes.py

```diff
@@ -97,6 +147,223 @@ def _configure_interface(self):
+
+        self._apply_tags(profile_name, pod)
+
+        # Also set the profile for the workload.
+        print('Setting profile %s on endpoint %s' %
+              (profile_name, endpoint.endpoint_id))
+        self._datastore_client.set_profiles_on_endpoint(
+            profile_name, endpoint_id=endpoint.endpoint_id
+        )
+        print('Finished configuring profile.')
+
+    def _container_add(self, container_name, ip, interface):
+        """
+        Add a container (on this host) to Calico networking with the given IP.
+
+        :param container_name: The name or ID of the container.
+        :param ip: An IPAddress object with the desired IP to assign.
+        :param interface: The name of the interface in the container.
+        """
+        # Obtain information from Docker Client and validate container state
+        info = self._get_container_info(container_name)
+        self._validate_container_add(info)
+
+        # Obtain ip pool from datastore client, validate pool, and assign
+        # ip address through IPAM Client
+        ip = IPAddress(ip)
+        self._assign_address_in_pool_IPAM(ip)
+
+        # Obtain next hops from datastore client, validate next hop, and
+        # create Endpoint object
+        container_id = info["Id"]
+        ep = self._endpoint_init(container_id, ip)
+
+        # Create the veth, move into the container namespace, add the IP and
+        # set up the default routes.
+        # Update Endpoint in felix
+        pid = info["State"]["Pid"]
+        ep = self._endpoint_complete(ep, pid, ip, interface)
+
+        # Let the caller know what endpoint was created.
+        return ep
+
+    def _container_remove(self, container_name):
+        """
+        Remove the indicated container on this host from Calico networking
+
+        :param container_name: THe name or partial id of the container
+        """
+        # Resolve the name to ID.
+        workload_id = self._get_container_id(container_name)
+
+        # Find the endpoint ID. We need this to find any ACL rules
+        try:
+            endpoint = self._datastore_client.get_endpoint(
+                hostname=hostname,
+                orchestrator_id=ORCHESTRATOR_ID,
+                workload_id=workload_id
+            )
+        except KeyError:
+            print("Container %s doesn't contain any endpoints" % container_name)
+            sys.exit(1)
+
+        # Remove any IP address assignments that this endpoint has
+        for net in endpoint.ipv4_nets | endpoint.ipv6_nets:
+            assert(net.size == 1)
+            ip = net.ip
+            self._unassign_address_in_pool_IPAM(ip)
+
+        # Remove the endpoint
+        netns.remove_veth(endpoint.name)
+
+        # Remove the container from the datastore.
+        self._datastore_client.remove_workload(hostname, ORCHESTRATOR_ID, workload_id)
+
+        print("Removed Calico interface from %s" % container_name)
+
+    def _validate_container_add(self, info):
```

Comment:
I might structure this differently.

I think the validation code is not something we want to stick in pycalico - this is just a nice to have, and might vary per-orchestrator.


I'd rename this to `_validate_container_info(self, info)` which validates that the container info is good.  The bit of validation that checks that the endpoint doesn't exist could probably just be inlined in _container_add.